### PR TITLE
Feature registration problem data

### PR DESCRIPTION
**Pull request by MaxMEllon**

### CODE REVIEW

File: app/controllers/admin/sessions_controller.rb [hunk 1]
```diff
@@ -29,7 +29,7 @@ def edit; end
 
   def show
     @q = @session.missions.ransack(params[:q])
-    @missions = @q.result&.page(params[:page] || 1)&.per(20)
+    @missions = @q.result&.page(params[:page] || 1).per(20)
```
Comment (MaxMEllon) [inline app/controllers/admin/sessions_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Lint/SafeNavigationChain: Do not chain ordinary method call after safe navigation operator.

File: spec/factories/problem_tests.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FactoryGirl.define do
```
Comment (hukurou-s) [inline spec/factories/problem_tests.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/factories/tests.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
```
Comment (hukurou-s) [inline spec/factories/tests.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/factories/tests.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  factory :test do
+    mission_id 1
+    test_name "MyString"
```
Comment (hukurou-s) [inline spec/factories/tests.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/factories/tests.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    test_command "MyString"
```
Comment (hukurou-s) [inline spec/factories/tests.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/models/test_spec.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (hukurou-s) [inline spec/models/test_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/models/problem_test_spec.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (hukurou-s) [inline spec/models/problem_test_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: app/models/mission.rb [hunk 8]
```diff
@@ -25,6 +26,18 @@ def java_files(absolute = false)
...
 
+  def test_names
+    json = File.read "#{local_repository}/tests.json"
+    JSON.parse(json).map { |i| i['name'] }
+  end
+
+  def test_commands
+    json = File.read "#{local_repository}/tests.json"
+    names = JSON.parse(json).map { |i| i['name'] }
+    commands = JSON.parse(json).map { |i| i['command']}
```
Comment (hukurou-s) [inline app/models/mission.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/SpaceInsideBlockBraces: Space missing inside }.

File: app/models/problem_test.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class ProblemTest < ApplicationRecord
```
Comment (hukurou-s) [inline app/models/problem_test.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: app/models/test.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+class Test < ApplicationRecord
```
Comment (hukurou-s) [inline app/models/test.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: app/controllers/admin/sessions_controller.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
+# coding: utf-8
```
Comment (hukurou-s) [inline app/controllers/admin/sessions_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/Encoding: Unnecessary utf-8 encoding comment.

File: app/controllers/admin/missions_controller.rb [hunk 12]
```diff
@@ -20,15 +20,17 @@ def show; end
...
     @mission.java_files.each do |file|
-      params[:locations].fetch(file, nil)&.permit!.to_h.map do |_key, value|
+      params[:locations].fetch(file, nil)&.permit!.to_h.map do |key, value|
         next unless value.respond_to?(:map)
         PiercedLocation.create(
           mission: @mission,
           lines: value.map(&:to_i),
-          file_name: file
+          file_name: file,
+          location_id: (key.to_i+1)
```
Comment (hukurou-s) [inline app/controllers/admin/missions_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/SpaceAroundOperators: Surrounding space missing for operator `+`.

File: app/controllers/admin/problems_controller.rb [hunk 13]
```diff
@@ -1,4 +1,57 @@
...
 class Admin::ProblemsController < Admin::ApplicationController
+  before_action :set_session
+  before_action :set_mission
+  before_action :set_problem, only: %i[show]
+
+  def new
+    @problem = @mission.problems.new
+  end
+
+  def create
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [29.27/15]

Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/MethodLength: Method has too many lines. [21/10]

File: app/controllers/admin/problems_controller.rb [hunk 14]
```diff
@@ -1,4 +1,57 @@
...
+    @problem = Problem.create(
+      mission: @mission,
+      name: params[:problem][:name],
+      detail: params[:problem][:detail],
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/TrailingCommaInArguments: Avoid comma after the last parameter of a method call.

File: app/controllers/admin/problems_controller.rb [hunk 15]
```diff
@@ -1,4 +1,57 @@
...
+    )
+
+    test_command_list =  @mission.test_commands
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/SpaceAroundOperators: Operator `=` should be surrounded by a single space.

Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/ExtraSpacing: Unnecessary spacing detected.

File: app/controllers/admin/problems_controller.rb [hunk 16]
```diff
@@ -1,4 +1,57 @@
...
+      )
+
+      ProblemTest.create(
+        problem_id: @problem.id,
+        test_id: test.id,
+        pierced_level: params[:labels][index]
+      )
+    end
+
+    redirect_to admin_session_mission_problems_path(params[:session_id], params[:mission_id], id: @problem.to_param)
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/LineLength: Line is too long. [116/100]

File: app/controllers/admin/problems_controller.rb [hunk 17]
```diff
@@ -1,4 +1,57 @@
...
+  end
+
+  def set_mission
+    @mission = Mission.find(params[:mission_id])
+  end
+
+  def set_session
+    @session = Session.find(params[:session_id])
+  end
+
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/EmptyLinesAroundClassBody: Extra empty line detected at class body end.

File: app/controllers/admin/problems_controller.rb [hunk 18]
```diff
@@ -1,4 +1,57 @@
...
 end
+
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/TrailingBlankLines: 1 trailing blank lines detected.

File: app/controllers/admin/problems_controller.rb [hunk 19]
```diff
@@ -1,4 +1,74 @@
...
 class Admin::ProblemsController < Admin::ApplicationController
+  before_action :set_session
+  before_action :set_mission
+  before_action :set_problem, only: %i[show]
+
+  def new
+    @problem = @mission.problems.new
+  end
+
+  def create
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [48.75/15]

Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/MethodLength: Method has too many lines. [33/10]

File: app/controllers/admin/problems_controller.rb [hunk 20]
```diff
@@ -1,4 +1,74 @@
...
+    @problem = Problem.new(problem_params)
+    unless @problem.save
+      logger.info problem_errors: @problem.errors.full_messages
+      flash[:alert] = "fail to create problem #{@problem.errors.full_messages.join(' ')}"
+      render :new
+    end
+
+    test_command_list = @mission.test_commands
+    tests = Test.where("mission_id = ?", params[:mission_id])
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: app/controllers/admin/problems_controller.rb [hunk 21]
```diff
@@ -1,4 +1,74 @@
...
+          test_command: test_command_list[test],
+          pierced_location_id: params['pirced-location-id'][index]
+        )
+      end
+
+      @problem_test = @test.problem_tests.build(
+        problem: @problem,
+        pierced_level: params[:labels][index]
+      )
+      unless @test.save
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/Next: Use `next` to skip iteration.

File: app/controllers/admin/problems_controller.rb [hunk 22]
```diff
@@ -1,4 +1,74 @@
...
+        logger.info test_errors: @test.errors.full_messages
+        flash[:alert] = "fail to create test #{@test.errors.full_messages.join(' ')}"
+        render :new
+      end
+    end
+
+    redirect_to admin_session_mission_problems_path(
+                  params[:session_id],
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/FirstParameterIndentation: Indent the first parameter one step more than the start of the previous line.

File: app/controllers/admin/problems_controller.rb [hunk 23]
```diff
@@ -1,4 +1,74 @@
...
+                  params[:mission_id],
+                  id: @problem.to_param
+                )
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/ClosingParenthesisIndentation: Indent `)` the same as the start of the line where `(` is.

File: app/controllers/admin/problems_controller.rb [hunk 24]
```diff
@@ -1,4 +1,70 @@
...
 class Admin::ProblemsController < Admin::ApplicationController
+  before_action :set_session
+  before_action :set_mission
+  before_action :set_problem, only: %i[show]
+
+  def new
+    @problem = @mission.problems.new
+  end
+
+  def create
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [42.86/15]

Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/MethodLength: Method has too many lines. [29/10]

File: app/javascript/components/admin/problem/ThePircedLocationStateEditor.js [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+`
+
+export default function ThePircedLocationStateEditor({ problem, testList }) {
+  return (
+    <Box>
+      <input type="text" value={problem.name} />
+      <div>
+        {problem.piercedLocationStateList.map(l => (
+          <Li key={Math.random()}>
+            <input type="hidden" name="pirced-location-id[]" value={l.id}/>
```
Comment (hukurou-s) [inline app/javascript/components/admin/problem/ThePircedLocationStateEditor.js]: **[eslint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Insert `·` (prettier/prettier)

File: app/javascript/components/admin/problem/ThePircedLocationStateEditor.js [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+              <div className="select">
+                <select name="labels[]">
+                  {Object.keys(PiercedLocationState).map(key => <option key={key}>{key}</option>)}
+                </select>
+              </div>
+            </Form1>
+            <Form2 className="control">
+              <label className="label is-small">対応テスト</label>
+              <div className="select">
+                <select name="tests[]">{testList.map(key => <option key={key}>{key}</option>)}</select>
```
Comment (hukurou-s) [inline app/javascript/components/admin/problem/ThePircedLocationStateEditor.js]: **[eslint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Replace `{testList.map(key·=>·<option·key={key}>{key}</option>)}` with `⏎··················{testList.map(key·=>·<option·key={key}>{key}</option>)}⏎················` (prettier/prettier)

File: app/javascript/components/admin/problem/TheProblemEditor.js [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+    const { piercedLocations, javaMainContents, testList } = this.props
+    this.props.initializeMetaOfProblemEditor({
+      piercedLocations,
+      javaMainContents,
+      testList
+    })
+  }
+
+  render() {
+    console.log(this.props.testList)
```
Comment (hukurou-s) [inline app/javascript/components/admin/problem/TheProblemEditor.js]: **[eslint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Unexpected console statement. (no-console)

File: app/controllers/admin/problems_controller.rb [hunk 28]
```diff
@@ -1,4 +1,47 @@
...
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [16.61/15]

File: app/javascript/components/admin/problem/TheProblemEditor.js (hunk 27 above)
Comment (hukurou-s) [inline app/javascript/components/admin/problem/TheProblemEditor.js]: **[eslint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Unexpected console statement. (no-console)

File: app/controllers/admin/problems_controller.rb [hunk 29]
```diff
@@ -1,4 +1,53 @@
...
+    problems = Problem.where('mission_id = ?', @mission.id).order('name')
+    @q = problems.ransack(params[:q])
+    @problems = @q.result&.page(params[:page] || 1)&.per(10)
+  end
+
+  def new
+    @problem = @mission.problems.new
+  end
+
+  def create
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [16.61/15]

File: app/controllers/admin/problems_controller.rb [hunk 30]
```diff
@@ -5,6 +5,12 @@ class Admin::ProblemsController < Admin::ApplicationController
   before_action :set_mission
   before_action :set_problem, only: %i[show]
 
+  def index
+    problems = Problem.where('mission_id = ?', @mission.id).order('name')
```
Comment (MaxMEllon) [inline app/controllers/admin/problems_controller.rb]:
```rb
Problem.where(mission_id: @mission_id)
```
のほうが良さそう

僕ならこうする

```rb
class Problem < Model
  ...
  scope :of_mission, -> (mission_id) { where(mission_id: mission_id) } 
  ...
end
```

```rb
class ProblemsController < Controller
  def index
    problems = Problem.of_mission(@mission.id).order(:name)
  end
end
```

File: app/controllers/admin/problems_controller.rb [hunk 31]
```diff
@@ -1,4 +1,64 @@
...
+    @q = problems.ransack(params[:q])
+    @problems = @q.result&.page(params[:page] || 1)&.per(10)
+  end
+
+  def new
+    @problem = @mission.problems.new
+  end
+
+  def create
```
Comment (hukurou-s) [inline app/controllers/admin/problems_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Metrics/AbcSize: Assignment Branch Condition size for create is too high. [16.61/15]

File: app/controllers/admin/problems_controller.rb (hunk 30 above)
Comment (hukurou-s) [reply app/controllers/admin/problems_controller.rb]: レビューいただきありがとうございます！
今気づきました。気づくの遅くてすみません...

ちょっと今調べて見たんですが、scope良さそうですね
他のところも含めて書き直してます！

File: spec/models/question_test_spec.rb [hunk 32]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (hukurou-s) [inline spec/models/question_test_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/models/question_spec.rb [hunk 33]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (hukurou-s) [inline spec/models/question_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/controllers/admin/questions_controller_spec.rb [hunk 34]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (hukurou-s) [inline spec/controllers/admin/questions_controller_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/controllers/admin/questions_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,5 @@
...
+
+RSpec.describe Admin::QuestionsController, type: :controller do
+
```
Comment (hukurou-s) [inline spec/controllers/admin/questions_controller_spec.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Layout/EmptyLinesAroundBlockBody: Extra empty line detected at block body beginning.

File: spec/factories/question_tests.rb [hunk 36]
```diff
@@ -0,0 +1,8 @@
+FactoryGirl.define do
```
Comment (hukurou-s) [inline spec/factories/question_tests.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/factories/questions.rb [hunk 37]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
```
Comment (hukurou-s) [inline spec/factories/questions.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: spec/factories/questions.rb [hunk 38]
```diff
@@ -0,0 +1,7 @@
...
+  factory :question do
+    problem_id 1
+    name "MyString"
```
Comment (hukurou-s) [inline spec/factories/questions.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: spec/factories/questions.rb [hunk 39]
```diff
@@ -0,0 +1,7 @@
...
+    detail "MyText"
```
Comment (hukurou-s) [inline spec/factories/questions.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: app/models/question_test.rb [hunk 40]
```diff
@@ -0,0 +1,5 @@
+class QuestionTest < ApplicationRecord
```
Comment (hukurou-s) [inline app/models/question_test.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: app/models/question.rb [hunk 41]
```diff
@@ -0,0 +1,5 @@
+class Question < ApplicationRecord
```
Comment (hukurou-s) [inline app/models/question.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

File: app/controllers/admin/questions_controller.rb [hunk 42]
```diff
@@ -0,0 +1,24 @@
+class Admin::QuestionsController < ApplicationController
```
Comment (hukurou-s) [inline app/controllers/admin/questions_controller.rb]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Style/FrozenStringLiteralComment: Missing magic comment `# frozen_string_literal: true`.

**Comment by MaxMEllon** [comment]
ついにマージされた

**Comment by hukurou-s** [comment]
これから加速していきます

**Pull request merged**